File: Code Wars Sort the Odd.py
input = [5, 3, 2, 8, 1, 4]
print("Input: ", input)


def sort_array(source_array):
    dict_odd_numbers = {}
    for index in range(len(source_array)):
        if (source_array[index] % 2) != 0:
            dict_odd_numbers[index] = source_array[index]
    """
    Try sort using both method and some sorting algorithm,
    probably Bubble Sort due to lack of information.
    """
    keys = list(dict_odd_numbers.keys())
    print("Odd keys' Indexes:", keys, sep='')

    while True:
        found_unordered = False

        for index in range(len(keys) - 1):
            one = dict_odd_numbers[keys[index]]
            two = dict_odd_numbers[keys[index + 1]]

            if one > two and one != 0 and two != 0:
                dict_odd_numbers[keys[index]] = two
                dict_odd_numbers[keys[index + 1]] = one
                found_unordered = True
        if not found_unordered:
            break
    # Return numbers in order to list.
    for index, value in dict_odd_numbers.items():
        source_array[index] = value

    return source_array


print("Output:", sort_array(input))

File: Code Wars Sort the Odd.jl


function sort_array(array)
    odd_numbers = []
    odd_number_indexes = []
    new_array = copy(array)

    for index in 1:length(array)
        number = array[index]
        if isodd(number)
            push!(odd_numbers, number)
            push!(odd_number_indexes, index)
        end
    end       
    
    while true
        ctr_unordered_pairs = 0

        for index in 1:(length(odd_numbers) - 1)
            first = odd_numbers[index]
            second = odd_numbers[index + 1]

            if first > second
                odd_numbers[index] = second
                odd_numbers[index + 1] = first
                ctr_unordered_pairs += 1
            end

        end

        if ctr_unordered_pairs == 0
            break
        end
    
    end

    counter = 1
        
    for index in odd_number_indexes
        new_array[index] = odd_numbers[counter]
        counter += 1
    end

    println("Odd indexes: ", odd_number_indexes)
    return new_array

end

"""
function sort_array_b(array)
    odd_numbers = Dict()
    new_array = copy(array)

    for index in 1:length(array)
        number = array[index]

        if isodd(number)
            push!(odd_numbers, index => number)            
        end
    end

    odd_number_indexes = keys(odd_numbers)

    ctr_unordered_pairs = 1

    while ctr_unordered_pairs >= 1
        ctr_unordered_pairs = 0

        for index in 1:(length(odd_number_indexes) - 1)
            first = odd_numbers[odd_number_indexes[index]]
            second = odd_numbers[odd_number_indexes[index + 1]]

            if first > second
                odd_numbers[odd_number_indexes[index]] = second
                odd_numbers[odd_number_indexes[index + 1]] = first
                ctr_unordered_pairs += 1
            end
        end

    end

    for odd_number_index in keys(odd_numbers)
        new_array[odd_number_index] = odd_numbers[odd_number_index]
    end

    println("Odd indexes: ", odd_number_indexes)
    return new_array
end
"""

function sort_array_c(array)
    odd_numbers = []
    odd_number_indexes = []
    new_array = copy(array)

    for index in 1:length(array)
        number = array[index]
        if isodd(number)
            push!(odd_numbers, number)
            push!(odd_number_indexes, index)
        end
    end       
    
    sort!(odd_numbers)

    counter = 1
        
    for index in odd_number_indexes
        new_array[index] = odd_numbers[counter]
        counter += 1
    end

    println("Odd indexes: ", odd_number_indexes)
    return new_array

end


input = [5, 3, 2, 8, 1, 4]
println("Input ", input)
println("Called sort_array: ", sort_array(input))
#println("Called sort_array_b: ", sort_array_b(input))
println("Called sort_array_c: ", sort_array_c(input))
